test/datasets/test_iwslt2016.py

```diff
@@ -0,0 +1,83 @@
+import os
+import random
+import string
+from collections import defaultdict
+from unittest.mock import patch
+
+from parameterized import parameterized
+from torchtext.datasets.iwslt2016 import IWSLT2016
+from torchtext.data.datasets_utils import _generate_iwslt_files_for_lang_and_split
+
+from ..common.case_utils import TempDirMixin, zip_equal
+from ..common.torchtext_test_case import TorchtextTestCase
+
+
+def _get_mock_dataset(root_dir, split, src, tgt):
+    """
+    root_dir: directory to the mocked dataset
+    """
+    temp_dataset_dir = os.path.join(root_dir, f"IWSLT2016/2016-01/texts/{src}/{tgt}/{src}-{tgt}/")
+    os.makedirs(temp_dataset_dir, exist_ok=True)
+
+    seed = 1
+    mocked_data = defaultdict(lambda: defaultdict(list))
+    valid_set = "tst2013"
+    test_set = "tst2014"
+
+    files_for_split, _ = _generate_iwslt_files_for_lang_and_split(16, src, tgt, valid_set, test_set)
+    src_file = files_for_split[src][split]
+    tgt_file = files_for_split[tgt][split]
+    for file_name in (src_file, tgt_file):
+        txt_file = os.path.join(temp_dataset_dir, file_name)
+        with open(txt_file, "w") as f:
+            # Get file extension (i.e., the language) without the . prefix (.en -> en)
+            lang = os.path.splitext(file_name)[1][1:]
+            for i in range(5):
+                rand_string = " ".join(
+                    random.choice(string.ascii_letters) for i in range(seed)
+                )
+                dataset_line = f"{rand_string} {rand_string}\n"
+                # append line to correct dataset split
+                mocked_data[split][lang].append(dataset_line)
+                f.write(f'{rand_string} {rand_string}\n')
+                seed += 1
+
+    return list(zip(mocked_data[split][src], mocked_data[split][tgt]))
```

Comment:
Is there a reason we are creating a download archive `2016-01.tgz` like we are doing for other datasets?

Reply:
> Is there a reason we are not creating a download archive `2016-01.tgz` like we are doing for other datasets?
> 
> Edit: I think it quite important to start from the download archive, otherwise we can get into hard to find bugs specially when the compression pattern is complex like we have in here.

@erip just wanted to check if you do plan to follow-up on this as well?

Reply:
Yes, I can follow up on this. It will take a lot more thought since, as you mention, the clean up is quite involved. That said, I think it should be doable.

Reply:
> Yes, I can follow up on this. It will take a lot more thought since, as you mention, the clean up is quite involved. That said, I think it should be doable.

Sure, thanks @erip!